### ISSUE TITLE
Private Tracker Support?

### ISSUE THREAD

**Issue by HDVinnie** [greatest-ape/aquatic#27]
Is private tracker support planned? Would be nice to implement in a project I maintain. 

https://github.com/HDInnovations/UNIT3D-Community-Edition

**Comment by shirshak55** [comment]
@HDVinnie which one do you use? udp or HTTP or ws ?

**Comment by HDVinnie** [comment]
HTTP

**Comment by greatest-ape** [comment]
aquatic is quite heavily focused on the open tracker use case. It keeps all swarm data in in-process memory, and additionally, locally to the threads accessing it when possible. A lot of the work that has gone into aquatic has been centered on finding out how to do this in a way that maximizes throughput.

A private tracker will want to persist data in a relational database. I assume that it might also want to perform access control for requests based on data in the database. In other words, you're looking at something much more similar to a typical back-end service: spawn a number of threads which receive HTTP requests, do a number of async calls to a DB and send back responses.

In other words, supporting private trackers is not compatible with _aquatic_http_ as it is currently structured. The only part that could be reused is the functionality for parsing and serializing requests and responses in _aquatic_http_protocol_ crate.

Additionally, I doubt that there is a way to implement private tracker support generic enough to support multiple implementations, so in practice, what would be needed is a custom new Rust tracker for UNIT3D that maybe uses some aquatic code.

Unfortunately, for the time being, I don’t have time to develop such an implementation.

**Comment by shirshak55** [comment]
@greatest-ape thanks. Are there any examples, spec related to the private tracker, etc..?

**Comment by greatest-ape** [comment]
> @greatest-ape thanks. Are there any examples, spec related to the private tracker, etc..?

Hmm, I don't know, maybe check out HDVinnie's link?

**Comment by HDVinnie** [comment]
* [BEP 27](http://bittorrent.org/beps/bep_0027.html): Private Torrents

**Comment by greatest-ape** [comment]
@HDVinnie Come to think of it, I think there might be a way to do this.

I'm thinking along the lines of having aquatic call a DB stored procedure on every announce request. It would return a boolean indicating if the request is accepted, i.e., the peer can download/upload the torrent, as well as an optional failure reason or warning message.

The actual contents of the stored procedure would be defined by the whole site surrounding the tracker (e.g. UNIT3D), in order to decouple from the implementation.

__Stored procedure name__

aquatic_announce_v1 (v1 to ease developing the protocol later)

__Stored procedure IN parameters__

| parameter name | mysql type           | notes                                                                                                                          |
|----------------|----------------------|--------------------------------------------------------------------------------------------------------------------------------|
| info_hash      | VARCHAR(20) NOT NULL |                                                                                                                                |
| user_token     | VARCHAR(40) NOT NULL | Would be included in announce URL and be unique per user                                                                       |
| peer_id        | VARCHAR(20) NOT NULL | Could be used to check that only whitelisted clients are used                                                                  |
| event          | VARCHAR(9) NOT NULL  | Useful for bookkeeping and cheating prevention                                                                                 |
| uploaded       | BIGINT NOT NULL      | Necessary for bookkeeping and cheating prevention                                                                              |
| downloaded     | BIGINT NOT NULL      | Necessary for bookkeeping and cheating prevention. Could also be used to determine if user is allowed to proceed with download |

__Stored procedure OUT parameters__

| parameter name   | mysql type       | notes    |
|------------------|------------------|----------|
| announce_allowed | BOOLEAN NOT NULL |          |
| failure_reason   | VARCHAR(255)     | optional |
| warning_message  | VARCHAR(255)     | optional |

aquatic would keep its own swarm state, while the rest of the tracker would be responsible for detecting cheating, managing ratios and so on.

What do you think? Is there anything else that would actually be needed from aquatic?

**Comment by erincerys** [comment]
You can look at ocelot's data layer implementation as a reference: https://github.com/WhatCD/Ocelot/blob/master/db.cpp

But since it was based on the extremely dated and known to be inefficient XBT, it might not be the greatest to replicate.

I like your minimal schema, tbh. Seems like it wouldn't run into some of those performance issues XBT did with recording too much information about snatches and swarm metadata.

I can't think of anything else off the top of my head that the tracker itself would absolutely need to be involved with. For example, when a user of the web UI its integrated with creates a new torrent, it could inspect the `.torrent` file to get the filesize to record in its own torrent metadata table that would also be used to set leech modifiers (freeleech, doubleseed).

As long as you're setting an integer column as the primary key on the aquatic-maintained table, a web frontend integrating with it would be able to reference that for table joins and efficient indexing.

(Full disclosure: while I've quite a bit of experience with data design and DB operations, my experience with tracker development and operations is limited comparatively.)

**Comment by HDVinnie** [comment]
Looks good....maybe `user_agent` might be good to have in there?

**Comment by greatest-ape** [comment]
Yeah, `user_agent` makes sense. I added it to the proposal.

I think p_downloaded and p_uploaded should be exactly what the peer reports, not a calculated change since last announce. The reason is that the stored procedure will have to check the DB to determine if it is a restart, fresh download etc.

A few more thoughts on my design:

- It has to hit the database on each announce, which will limit performance
  - On the other hand, even if DB accesses are batched like in Ocelot, a query will have to be executed for each announce request anyway, which might not be much faster
  - Load could maybe be reduced to some extent by also keeping lists of all allowed user_tokens and info_hashes (updated say every 10 seconds) and filtering by them before checking with the DB. This would mostly make a difference for malicious requests, though, and might be considered a premature optimisation at this point.
- Business logic has to be written in SQL (unless stored procedure mostly just stores data and processes it later). This might be unwieldy and means higher DB load than doing calculations in aquatic. However, I don't see another way to keep aquatic decoupled.

@HDVinnie do you have an up to date structure-only UNIT3D DB dump that I could use for testing?

@erincerys Thanks for the feedback. To clarify, I'm not thinking of having an aquatic-maintained table. The stored procedure would be responsible to doing what it wants to do with data it receives (storing it etc) according to internal requirements.

**Comment by HDVinnie** [comment]
[![Screen-Shot-2022-03-10-at-3-35-43-PM.png](https://i.postimg.cc/pdGhtzbw/Screen-Shot-2022-03-10-at-3-35-43-PM.png)](https://postimg.cc/yJRYmJNj)

Here's schema dump from latest UNIT3D.
https://anonfiles.com/Tc95BcSbxc/mysql-schema_dump

**Comment by greatest-ape** [comment]
Closing due to inactivity.
